presto-orc/src/main/java/com/facebook/presto/orc/reader/FloatSelectiveStreamReader.java

```diff
@@ -0,0 +1,403 @@
+        }
+
+        return positions[positionCount - 1] + 1;
+    }
+
+    private int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        // filter == null implies outputRequired == true
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                nulls[i] = true;
+            }
+            else {
+                values[i] = floatToRawIntBits(dataStream.next());
+                if (presentStream != null) {
+                    nulls[i] = false;
+                }
+            }
+            streamPosition++;
+        }
+        outputPositionCount = positionCount;
+        return streamPosition;
+    }
+
+    private void skip(int items)
+            throws IOException
+    {
+        if (dataStream == null) {
+            presentStream.skip(items);
+        }
+        else if (presentStream != null) {
+            int dataToSkip = presentStream.countBitsSet(items);
+            dataStream.skip(dataToSkip);
+        }
+        else {
+            dataStream.skip(items);
+        }
+    }
+
+    private void ensureValuesCapacity(int capacity, boolean recordNulls)
+    {
+        if (values == null || values.length < capacity) {
+            values = new long[capacity];
+        }
+
+        if (recordNulls) {
+            if (nulls == null || nulls.length < capacity) {
+                nulls = new boolean[capacity];
+            }
+        }
+    }
+
+    private void ensureOutputPositionsCapacity(int capacity)
+    {
+        if (outputPositions == null || outputPositions.length < capacity) {
+            outputPositions = new int[capacity];
+        }
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(NULL_BLOCK, outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        if (positionCount == outputPositionCount) {
+            Block block = new LongArrayBlock(positionCount, Optional.ofNullable(includeNulls ? nulls : null), values);
+            nulls = null;
+            values = null;
+            return block;
+        }
+
+        long[] valuesCopy = new long[positionCount];
```

Comment:
How do you feel about extracting this logic into a utility class and re-using it across readers that produce LongArrayBlocks (e.g. long, timestamp and float)?